pkg/transform/README.md

```diff
@@ -0,0 +1,131 @@
+![The Buf logo](../../.github/buf-logo.svg)
+
+# Transform
+
+[![Slack](https://img.shields.io/badge/slack-buf-%23e01563)][badges_slack]
+
+_This project is currently in **alpha**. The API should be considered unstable and likely to change._
+
+Allows the caller to convert a given message data blob from one format to
+another by referring to a type schema for the blob.
+
+A default client provides you with connection to the public Buf Schema Registry
+and an in memory cache which will maintain the `protoencoding.Resolver` for
+future transformations. You must provide it with a reference to a Buf Module and
+the expected format of the data you wish to convert and the resulting format you
+would like.
+
+```go
+package foo
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/bufbuild/buf/pkg/transform"
+	registryv1alpha1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/registry/v1alpha1"
```

Comment:
External code won't be able to import this since it's under `private`.

I just realized something much worse: it's not just that external code can't _import_ private packages, they also can't even be _linked_ to private packages. In other words, this `pkg/transform` cannot use any private packages, _even indirectly_, or else external binaries that try to use it will [blow up](https://github.com/bufbuild/buf/blob/main/private/usage/usage.go#L28-L32) during `init`.

So I think we have two options with how to proceed:
1. Remove all references to the private packages.
    * Don't use these types in this package's API. Create a separate enum type for the format, for example.
    * Add a step to `make generate` that generates only the schema service (can use the new type filtering stuff in `buf generate`) into some place like `pkg/transform/internal`).
    * We'd likely need a CI check that can make sure we never accidentally add a transitive dependency from pkg -> private since that would break users.
2. Move this package under private and punt on making it a library for external users.
    * Basically, we try to incorporate it into things only as an internal library first (like a sub-command of the buf CLI or a stand-alone executable that we provide; anything that lives under "github.com/bufbuild" is allowed to import the private packages.).

I think we may find the second option the path of least resistance for now. And we will need to have some serious conversations about how we can create open-source libraries. (This inevitably will involve pulling some stuff out of private; we'd likely want to mostly move it to internal packages, so that they don't become part of exported API that we have to support.)